### DISCUSSION TITLE
Fixing the particle number of a tensor basis using the quspin package

### DISCUSSION THREAD

**Discussion by ToastyTomatos** [QuSpin/QuSpin#693] [Q&A] [answered]
We want to describe a system with a fixed number of particles. We model our system using the quspin package (https://quspin.github.io/QuSpin/index.html). The problem is that we haven't found a way to fix the total number of particles in our combined basis defined as "basis":

Lferm=8 
Nbosons=20

basis_fermion=spinless_fermion_basis_1d(L=Lferm, Nf=range(0,Lferm+1,2))

basis_boson = boson_basis_1d(L=1, sps=Nbosons+1, Nb=range(0,Nbosons+1))

basis = tensor_basis(basis_fermion, basis_boson)

While it is easy to find all combinations of fermionic & bosonic basis vectors that have a total particle number of f.e. 10, and extract all the parameters that identify these basis vectors (f.e. index, integer representation etc.), we have not been able to redefine our basis using this information, i.e. we cannot generate a new basis that is still of the "quspin basis type" containing only basis vectors with a fixed number of particles.

Does anybody know a solution/workaround to this problem **WITHOUT** using user_basis ?  

PS: While digging through the package, I have found the sentence:

"The `tensor_basis` class does not allow one to make use of symmetries, save for particle conservation."

which makes me think there might be an easy way to implement the particle number fixing of the tensor_basis

**Comment by mgbukov** [comment, marked as answer]
one way you should be able to do this is using the user_basis.

This is a bit involved but there's a detailed tutorial (+examples) on the
website on how to use it. There are example codes for bosons and fermions
in the user_basis separately: what you'd have to do is figure out how to
combine them.

The benefit of learning how to use the user_basis is that then you will be
able to do many more cool things with quspin.
Marín

**Reply by ToastyTomatos** [reply]
Thanks for the quick answer! 
I wanted to avoid having to use user_basis due to time constraints of the project. Just to be clear, this means there is no way besides user_basis to implement this?

**Comment by mgbukov** [comment]
I don't think there is...
Marín
